File: leetcode/algorithms/118_pascals_triangle/main.dart
class PascalsTriangle {
  /**
   * DP
   * 
   * Complexities:
   *   N - `numRows`
   *   - Time Complexity: O(N²)
   *   - Space Complexity: O(N²)
   */
  List<List<int>> generate(int numRows) {
    List<List<int>> result = List<List<int>>.empty(growable: true);

    for (var i = 0; i < numRows; i++) {
      List<int> row = List<int>.empty(growable: true);
      for (var j = 0; j <= i; j++) {
        if (j == 0 || j == i) {
          row.add(1);
        } else {
          row.add(result[i - 1][j - 1] + result[i - 1][j]);
        }
      }
      result.add(row);
    }

    return result;
  }


  // Solution
  /**
   * DP
   * 
   * Complexities:
   *   N - `numRows`
   *   - Time Complexity: O(N²)
   *   - Space Complexity: O(N²)
   */
  List<List<int>> solution(int numRows) {
    if (numRows <= 0) {
      return [];
    }

    List<List<int>> triangle = [];

    for (int i = 0; i < numRows; i++) {
      List<int> row = List.filled(i + 1, 1);

      for (int j = 1; j < i; j++) {
        row[j] = triangle[i - 1][j - 1] + triangle[i - 1][j];
      }

      triangle.add(row);
    }

    return triangle;
  }
}

File: leetcode/algorithms/118_pascals_triangle/main.java
import java.util.ArrayList;
import java.util.List;

class PascalsTriangle {
  /**
   * DP
   * 
   * Complexities:
   *   N - `numRows`
   *   - Time Complexity: O(N²)
   *   - Space Complexity: O(N²)
   */
  public List<List<Integer>> generate(int numRows) {
    List<List<Integer>> result = new ArrayList<>();

    for (int i = 0; i < numRows; i++) {
      List<Integer> row = new ArrayList<>();

      for (int j = 0; j <= i; j++) {
        if (j == 0 || j == i) {
          row.add(1);
        } else {
          row.add(result.get(i - 1).get(j - 1) + result.get(i - 1).get(j));
        }
      }

      result.add(row);
    }

    return result;
  }


  // Solution
  /**
   * DP
   * 
   * Complexities:
   *   N - `numRows`
   *   - Time Complexity: O(N²)
   *   - Space Complexity: O(N²)
   */
  public List<List<Integer>> solution(int numRows) {
        List<List<Integer>> triangle = new ArrayList<>();

        if (numRows == 0) {
            return triangle;
        }

        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();

            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    row.add(1);
                } else {
                    List<Integer> prevRow = triangle.get(i - 1);
                    int sum = prevRow.get(j - 1) + prevRow.get(j);
                    row.add(sum);
                }
            }

            triangle.add(row);
        }

        return triangle;
  }
}
